### ISSUE TITLE
Build cannot generate files for enum and/or custom types (freezed/json_serializable)

### ISSUE THREAD

**Issue by SalahAdDin** [rrousselGit/freezed#652]
**Describe the bug**
Recently, I update `freezed` from `1.1.1` to `2.0.2` and at generating files I'm getting the current bug:

```bash
[SEVERE] json_serializable:json_serializable on lib/core/domain/settings/settings.entity.dart:

Could not generate `toJson` code for `introductoryVideo` because of type `IntroductoryVideo`.
package:thesis_cancer/core/domain/settings/settings.entity.freezed.dart:149:31
    ╷
149 │   List<IntroductoryVideo> get introductoryVideo {
    │                               ^^^^^^^^^^^^^^^^^
    ╵
package:json_serializable/src/encoder_helper.dart 136:7     EncodeHelper._serializeField
package:json_serializable/src/encoder_helper.dart 93:26     EncodeHelper._writeToJsonWithNullChecks
package:json_serializable/src/encoder_helper.dart 48:7      EncodeHelper.createToJson
dart:_internal                                              WhereIterator.moveNext
package:json_serializable/src/json_part_builder.dart 64:27  _UnifiedGenerator.generate
package:source_gen/src/builder.dart 352:23                  _generate

[FINE] json_serializable:json_serializable on lib/core/domain/settings/schedules.entity.dart:Running JsonSerializableGenerator - 1 of 2

```

The problem is obviously files for `IntroductoryVideo` were not generated:
![image](https://user-images.githubusercontent.com/5034215/164771831-b87ee74e-e850-41c3-9c65-20ea7a51956e.png)

Now, the code for this entity is:
```dart
part 'introvideo.entity.freezed.dart';
part 'introvideo.entity.g.dart';

/// [Introductory] Video Entity
@freezed
class IntroductoryVideo with _$IntroductoryVideo {
  /// Default builder
  const factory IntroductoryVideo({
    required UserRole role,
    required UploadFile video,
  }) = _IntroductoryVideo;

  /// Builds a [IntroductoryVideo] from a given json [String].
  factory IntroductoryVideo.fromJson(Map<String, Object?> json) =>
      _$IntroductoryVideoFromJson(json);

  /// Builds a empty(dummy) [IntroductoryVideo].
  static const IntroductoryVideo empty = IntroductoryVideo(
    role: UserRole.GUEST,
    video: UploadFile.empty,
  );
}
```

Where `enum UserRole { ADMIN, PILOT, TEST, CONTROL, GUEST }` and files for `UploadFile` are also generated:
![image](https://user-images.githubusercontent.com/5034215/164772424-f5e4de01-3ebc-4df1-bf64-61a4c3fc9e41.png)

Other classes with `enum` as type of one of the properties are getting the same problem:
```dart
part 'activityfeed.entity.freezed.dart';
part 'activityfeed.entity.g.dart';

/// Activity Feed (notification) Entity
@freezed
class ActivityFeed with _$ActivityFeed {
  /// Default constructor
  const factory ActivityFeed({
    String? id,
    required ActivityType type,
    @Default(<String, dynamic>{}) Map<String, dynamic>? data,
    required String body,
    required String title,
    @Default(false) bool isRead,
  }) = _ActivityFeed;

  const ActivityFeed._();

  /// Builds an [ActivityFeed] from a given [String] payload.
  factory ActivityFeed.fromPayload({required String payload}) {
    final Map<String, dynamic> data =
        jsonDecode(payload) as Map<String, Object?>;
    final ActivityFeed result = ActivityFeed.fromJson(data);
    return result;
  }

  /// Builds an [ActivityFeed] from a given json [String].
  factory ActivityFeed.fromJson(Map<String, Object?> json) =>
      _$ActivityFeedFromJson(json);

  /// Builds a empty(dummy) [ActivityFeed].
  static const ActivityFeed empty = ActivityFeed(
    type: ActivityType.NEW_POST,
    body: 'New dummy feed',
    title: 'Dummy Feed',
  );
}
```
![image](https://user-images.githubusercontent.com/5034215/164772720-1d0c1027-f69a-436c-a102-2b0be9f2efcd.png)

But some others without `enum` variables are also getting this problem:
```dart
part 'answer.entity.freezed.dart';
part 'answer.entity.g.dart';

///
@freezed
class UserSurveyAnswer with _$UserSurveyAnswer {
  ///
  const factory UserSurveyAnswer({
    // String? id,
    required String statement,
    required String answer,
  }) = _UserSurveyAnswer;

  ///
  factory UserSurveyAnswer.fromJson(Map<String, Object?> json) =>
      _$UserSurveyAnswerFromJson(json);

  ///
  static const UserSurveyAnswer empty = UserSurveyAnswer(
    statement: '',
    answer: '',
  );
}
```

Hence, I deduce there is a problem with the way they are being parsed by `freezed` for the version `2.0.0` and on. This bug is presents from `2.0.0` version and so on, which is a big regression.

This is my built settings:
```yaml
targets:
  $default:
    builders:
      json_serializable:
        options:
          explicit_to_json: true
          include_if_null: false
```

**To Reproduce**
```
import 'dart:convert';

import 'package:freezed_annotation/freezed_annotation.dart';

/// Notification ([ActivityFeed]) type.
enum ActivityType {
  // NEW_COMMENT,
  // NEW_FOLLOW,
  // NEW_LIKE,
  NEW_POST,
  SCHEDULED_SURVEY_REMINDER,
  // NEW_RECOMMENDATION,
  NEW_USER_REGISTERED,
  SETTINGS_UPDATED,
  USER_CONFIRMED
}



part 'activityfeed.entity.freezed.dart';
part 'activityfeed.entity.g.dart';

/// Activity Feed (notification) Entity
@freezed
class ActivityFeed with _$ActivityFeed {
  /// Default constructor
  const factory ActivityFeed({
    String? id,
    required ActivityType type,
    @Default(<String, dynamic>{}) Map<String, dynamic>? data,
    required String body,
    required String title,
    @Default(false) bool isRead,
  }) = _ActivityFeed;

  const ActivityFeed._();

  /// Builds an [ActivityFeed] from a given [String] payload.
  factory ActivityFeed.fromPayload({required String payload}) {
    final Map<String, dynamic> data =
        jsonDecode(payload) as Map<String, Object?>;
    final ActivityFeed result = ActivityFeed.fromJson(data);
    return result;
  }

  /// Builds an [ActivityFeed] from a given json [String].
  factory ActivityFeed.fromJson(Map<String, Object?> json) =>
      _$ActivityFeedFromJson(json);

  /// Builds a empty(dummy) [ActivityFeed].
  static const ActivityFeed empty = ActivityFeed(
    type: ActivityType.NEW_POST,
    body: 'New dummy feed',
    title: 'Dummy Feed',
  );
}
```

Execute `flutter pub run build_runner build --delete-conflicting-outputs --verbose`.

**Expected behavior**
It should generate all required files.

**Setup**
```
  freezed_annotation: ^2.0.1
  json_annotation: ^4.4.0
dev_dependencies:
  build_runner: ^2.1.10
  flutter_test:
    sdk: flutter
  freezed: ^2.0.0
  json_serializable: ^6.1.6
  lint: ^1.8.2
```

```
 flutter doctor -v
[✓] Flutter (Channel stable, 2.10.5, on Manjaro Linux 5.15.32-1-MANJARO, locale en_GB.UTF-8)
    • Flutter version 2.10.5 at /opt/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5464c5bac7 (hace 4 días), 2022-04-18 09:55:37 -0700
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /home/mehmeth/Android/Sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /usr/bin/chromium

[✓] Android Studio (version 2021.1)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 66.0.1
    • Dart plugin version 211.7817
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] Connected device (2 available)
    • MI 9 (mobile) • da22004 • android-arm64  • Android 10 (API 29)
    • Chrome (web)  • chrome  • web-javascript • Chromium 100.0.4896.127 Arch Linux

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

**Comment by gysipos** [comment]
Seems like i have the same issue here. On latest freezed and freezed_annotation, any time i run build_runner build/watch i lose a bunch of *.freezed.dart and *.g.dart files.
As a workaround i can comment out the @freezed annotation ( //@freezed ) in a file, trigger the build_runner watch, then de-comment the @freezed annotation (import freezed_annotation) and trigger the build_runner watch once again.

This time it generates the .freezed.dart and the .g.dart files correctly. As you can imagine, with 200+ freezed classes in my project, this is a long and tedious process, and offers only a temporary solution. A day and a few changes later running build_runner watch/build screws the freezed classes up again.

**Comment by hacker1024** [comment]
Yep, same here.
It seems that .freezed.dart and .g.dart files aren't generating at random, and even benign changes to files like adding comments will change which generated files are missing.

**Comment by hacker1024** [comment]
@SalahAdDin Perhaps you could mention that this issue was introduced in v2.0.0 to emphasize that this is a significant regression?

**Comment by wickyou23** [comment]
I got the same error when generating files with v2.0.0 or later. But when I downgrade to version v1.1.1, It works well.

**Comment by rrousselGit** [comment]
I don't have any problem with the example given, so I'm not sure what the issue is.

If you could isolate the problem and maybe create a github repository to reproduce it, that'd help a lot

**Comment by wickyou23** [comment]
I tried to create a new project and add models of my project, It work well. But in my project, if I keep models in an old folder, the build_runner detects missing some files that are marked ` @freezed`, see the picture, It just detected 6 files.

![Screen Shot 2022-05-10 at 14 16 42](https://user-images.githubusercontent.com/5183190/167582040-2199e2c1-5d82-48c4-b029-f1c8c76dce6c.png)

I tried to move models to other folders, and the build_runner worked well again.

![Screen Shot 2022-05-10 at 14 17 34](https://user-images.githubusercontent.com/5183190/167582068-83ac320f-dbb4-4bc6-b4ba-250c2304d30a.png)

@rrousselGit  What are the problems? I tried to run commands `flutter clean`, `flutter pub get`, `build_runner clean`, and `build_runner build` again, but It can not work.

**Comment by rrousselGit** [comment]
Like I said, I can't reproduce the issue. As such I don't know what the issue is.

**Comment by wickyou23** [comment]
> Like I said, I can't reproduce the issue. As such I don't know what the issue is.

Sorry, I can't find the root cause as well. But I have some information for you can investigate. If I remove `import` from all files that imported "freezed" model, It works well again.

**Comment by xxmn** [comment]
Yep, same here too.
move to another folder and it will be generated normally

**Comment by wickyou23** [comment]
> Yep, same here too. move to another folder and it will be generated normally

I think because when you move to another folder, the files `import` `freeze models` before will be invalid, so It generates normally.

**Comment by gysipos** [comment]
I could't be able to pin point the issue either. I tired to make a minimal repo and mess with some enums, but freezed worked like a champ :)
My hunch is that the issue is not in the code, but somewhere around it.

**Comment by SalahAdDin** [comment]
> > Yep, same here too. move to another folder and it will be generated normally
> 
> I think because when you move to another folder, the files `import` `freeze models` before will be invalid, so It generates normally.

Did you try removing the cache before to build everything?

**Comment by wickyou23** [comment]
> > > Yep, same here too. move to another folder and it will be generated normally
> > 
> > 
> > I think because when you move to another folder, the files `import` `freeze models` before will be invalid, so It generates normally.
> 
> Did you try removing the cache before to build everything?
Yes, I used:
`flutter clean`
`flutter pub get`
`flutter packages pub run build_runner clean`
`flutter packages pub run build_runner build --delete-conflicting-outputs`

But it does not work. :(

**Comment by wickyou23** [comment]
> I could't be able to pin point the issue either. I tired to make a minimal repo and mess with some enums, but freezed worked like a champ :) My hunch is that the issue is not in the code, but somewhere around it.

I don't think so. If the issue is not in the code, Why does v1.1.1 work normally? Currently, I'm using v1.1.1 to resolve this issue.

**Comment by gysipos** [comment]
Good point. What i meant is that it's not the generation itself, but the infrastructural code around it. After a bit of nudge (in case of a more complex file with a lot of freezed classes a lot of nudge) the generated code is perfect. So i think the issue lays around the actual code generation, somewhere where it sets itself up to generate code.

**Comment by rrousselGit** [comment]
I'd still like if someone could make a git repository to reproduce this issue. I doubt I'll be able to fix this problem otherwise.

**Comment by SalahAdDin** [comment]
> > > > Yep, same here too. move to another folder and it will be generated normally
> > > 
> > > 
> > > I think because when you move to another folder, the files `import` `freeze models` before will be invalid, so It generates normally.
> > 
> > 
> > Did you try removing the cache before to build everything?
> 
> Yes, I used: `flutter clean` `flutter pub get` `flutter packages pub run build_runner clean` `flutter packages pub run build_runner build --delete-conflicting-outputs`
> 
> But it does not work. :(

Yeah, that's my point, even if you clean the cache, it does not work.

**Comment by JonasWanke** [comment]
We also encountered this problem in our project and resolved it by downgrading to these versions:

```yaml
dependencies:
  analyzer: 3.4.1
  build: 2.2.1
  build_resolvers: 2.0.6
  freezed: 2.0.2
```

The problems did previously occur with the following versions:

```yaml
dependencies:
  analyzer: 4.0.0
  build: 2.3.0
  build_resolvers: 2.0.8
  freezed: 2.0.3+1
```

Hence, I don't think this is a problem with <kbd>freezed</kbd>, but rather with one of the other packages because v2.0.2 works for us while it leads to problems for others. However, I have not investigated which of the other dependencies is the root cause.

**Comment by hacker1024** [comment]
Interesting. It looks like the only build-system-related change in `freezed`'s dependencies between v1.1.1 and v2.0.0 is the version constraint on the `analyzer` package to include v4.x (https://github.com/rrousselGit/freezed/commit/d09037183d866ba387ced8ceadcdf39a42f0759f).

I don't think `build` or `build_resolvers` are transitive dependencies of the `analyzer` package, so the problem is likely in there.

**Comment by rrousselGit** [comment]
It's highly likely that the problem is in transitive dependencies indeed. I'm investigating this issue today, but I don't see why Freezed would have anything to do with this.

**Comment by wickyou23** [comment]
@rrousselGit My problem was resolved when upgrading the analyzer to 4.1.0. Thanks.

**Comment by SalahAdDin** [comment]
> @rrousselGit My problem was resolved when upgrading the analyzer to 4.1.0. Thanks.

Analyzer is still a transitive dependency.

**Comment by rrousselGit** [comment]
Since the problem seems to be resolved, I'll close this.
